mongo/integration/server_selection_prose_test.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package integration
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/event"
+	"go.mongodb.org/mongo-driver/mongo/integration/mtest"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// TestServerSelectionProse implements the Server Selection prose tests:
+// https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection-tests.rst
+func TestServerSelectionProse(t *testing.T) {
+	mt := mtest.New(t, mtest.NewOptions().CreateClient(false))
+	defer mt.Close()
+
+	mtOpts := mtest.NewOptions().Topologies(mtest.Sharded).MinServerVersion("4.9")
+	mt.RunOpts("operationCount-based selection within latency window, with failpoint", mtOpts, func(mt *mtest.T) {
+		_, err := mt.Coll.InsertOne(context.Background(), bson.D{})
+		require.NoError(mt, err, "InsertOne() error")
+
+		hosts := options.Client().ApplyURI(mtest.ClusterURI()).Hosts
+		require.GreaterOrEqualf(mt, len(hosts), 2, "test cluster must have at least 2 mongos hosts")
```

Comment:
Should this be a skip not a failure? I do think all the sharded configurations setup through mongo orchestration have two mongos hosts, though.

Reply:
Skipping is an option, although I'm concerned that our test environment could change and silently skip these tests. Since the test already only runs on sharded topologies, I'd prefer to get a test failure if the sharded test environment doesn't meet the test requirements.

Reply:
Sounds good.